### ISSUE TITLE
ArgumentException when 2.x version of NUnit Framework is in the bin directory

### ISSUE THREAD

**Issue by alex-mazzariol** [nunit/nunit#1018]
When testing an assembly with a target framework version which has only 2 fields (say, major = 4 and minor = 5, which is what I am testing now, compiled with mono 4.0.4) the code in NUnit.Engine.Services.TestAgency.LaunchAgentProcess() fails while calling System.Version.ToString(), raising an ArgumentException with message "Argument must be between 0 and 2. Parameter name: fieldCount".

I believe the problem to be around lines 204 and 204 of file  nunit/src/NUnitEngine/nunit.engine/Services/TestAgency.cs as of commit 272eadd. There the version is supposed to have at least 3 components, when there may not be that many.

**Comment by rprouse** [comment]
Which version of NUnit are you using and with what command line? We compile and on Linux with Mono, so we must be missing something. Can you provide more steps to reproduce?

**Comment by alex-mazzariol** [comment]
Just updated everything to the last 3.0 RC; version of NUnit framework and console runner is 3.0.5790.16288, command line is just
`nunit3-console.exe --noc TestAssembly.dll`
Assembly compiled with MonoDevelop's xbuild engine version 12.0, with Mono Version 4.0.4.0 (from Xamarin's RPM repository, running on Fedora 21). All the assemblies in the solution target "Mono / .NET 4.5" (from Project Options).

Problem persists also with this last version; when starting the console runner, this is the output:
[console-runner-20151112.txt](https://github.com/nunit/nunit/files/33348/console-runner-20151112.txt)

**Comment by CharliePoole** [comment]
The error would have to be somewhat upstream. We'll have to figure out how the ClrVersion field of the RuntimeFramework object ends up with only two digits. Do you know what this version of mono puts into the assembly as the image runtime version?

**Comment by alex-mazzariol** [comment]
Thank you for your fast responses; I checked the "image runtime version" using ILDASM on Windows, looking for "Metadata Version" as per http://blogs.msdn.com/b/suzcook/archive/2003/06/20/57191.aspx, however she says that the ImageRuntimeVersion of an Assembly may not be the actual required one.

ILDASM says the Metadata Version is 4.0.30319.

I also noticed that MonoDevelop did not update the DLL files (namely, nunit.framework.dll) in the output directory, so it kept an old version - 2.6.2 from net-40 directory. Once manually cleared the directory and rebuilt the solution, now the nunit.framework.dll version is in sync with the console runner, 3.0.5790, and is taken from the net-45 directory, and the problem disappears: the tests run correctly.

I don't think it is still a bug, when the assembly under test is linked to an old version (targeting a different framework, net-40) and the runner is new (and it crashes), so I think this issue can be closed - still, if you want to investigate on this problem, I'll try to help more.

**Comment by rprouse** [comment]
Are you saying that your tests referenced the 3.0 nunit framework, but the 2.6.2 version of the framework was in the output directory?

**Comment by alex-mazzariol** [comment]
Yes sir; due to a non-updated nunit.framework.dll in the output directory, the Assembly Under Test was referencing the 2.6.2 (net-40) framework, while my console runner was version 3. My bad for not noticing this before.

**Comment by rprouse** [comment]
@alex-mazzariol, thanks for clarifying. I am going to leave this open. We shouldn't be crashing, we should be displaying an error. If this is not a mono issue, but just an older framework in the bin directory issue, then it could happen more often.

I am mainly thinking of the scenario where people build multiple projects into the same bin directory. If some test projects have been upgraded, but not others, the framework could be any version. Of course, this is a mistake on the developers part as all dependencies for the solution should be at the same version, but ideally we wouldn't crash.

I am going to leave the confirm label on this to see if it is a more general non-Mono problem, but drop the priority to low since it is a user error and move it out to post 3.0. @CharliePoole, if you think we should look at this earlier, move it back :smile:

**Comment by alex-mazzariol** [comment]
Another thing I noticed: the same error occurs if the file name of the assembly under test does not exist. It may be related, i.e. some errors on the conflicting versions prevent an existing file from loading.

**Comment by CharliePoole** [comment]
It's a pretty strange error. For one thing, if you have the 2.6.2 framework instead of the 3.0 framework, then your tests should fail to load with an error that indicates there is an assembly mismatch.

I wonder if this is one of those situations where an exception is thrown while trying to display the error message, thereby masking the original error?

That said, I think it's a low priority. When we create milestones for bug fixes we can consider this for promotion to 3.0.1, 3.0.2, etc.

**Comment by mcraiha** [comment]
I am having the same issue. Running
**mono NUnit.Console.3.0.1/tools/nunit3-console.exe myproject/nunit-lib.csproj --config=Debug**
results
> NUnit Console Runner 3.0.5813 
> Copyright (C) 2015 Charlie Poole
> 
> Runtime Environment
>    OS Version: Linux 4.1.10.17 
>   CLR Version: 4.6.57.0
> 
> Test Files
>     myproject/nunit-lib.csproj
> 
> System.ArgumentException: Argument must be between 0 and 2.
> Parameter name: fieldCount
  at System.Version.ToString (Int32 fieldCount) <0x7fc3148c4650 + 0x00325> in <filename unknown>:0 
  at NUnit.Engine.Services.TestAgency.LaunchAgentProcess (NUnit.Engine.TestPackage package) <0x4151dbb0 + 0x0068b> in <filename unknown>:0 
  at NUnit.Engine.Services.TestAgency.CreateRemoteAgent (NUnit.Engine.TestPackage package, Int32 waitTime) <0x4151d980 + 0x00048> in <filename unknown>:0 
  at NUnit.Engine.Services.TestAgency.GetAgent (NUnit.Engine.TestPackage package, Int32 waitTime) <0x4151d950 + 0x0001e> in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) NUnit.Engine.Services.TestAgency:GetAgent (NUnit.Engine.TestPackage,int)
  at NUnit.Engine.Runners.ProcessRunner.LoadPackage () <0x4151d2f0 + 0x00136> in <filename unknown>:0 

nunit-lib.csproj can be found from http://pastebin.com/gpHsSP9i and TestCases.cs from http://pastebin.com/i8rkuk6A

Does someone have any tip how to avoid this issue?

**Comment by rprouse** [comment]
My best guess is that whatever you are using to compile is ignoring the 3.0.1 version of NUnit in your hint path and pulling the old version from the GAC.

- What distro of Linux?
- What version of Mono?
- Are you compiling command line or using MonoDevelop?
- How did you adding NUnit? It looks like  you are just browsing to it?

When I add a reference in Visual Studio, it adds the fully qualified nunit framework.

```XML
    <Reference Include="nunit.framework, Version=3.0.5813.39031, Culture=neutral, PublicKeyToken=2638cd05610744eb, processorArchitecture=MSIL">
      <HintPath>..\packages\NUnit.3.0.1\lib\net45\nunit.framework.dll</HintPath>
      <Private>True</Private>
    </Reference>
```

Whereas yours looks like this.

```XML
    <Reference Include="nunit.framework">
      <HintPath>..\NUnit.3.0.1\lib\net45\nunit.framework.dll</HintPath>
    </Reference>
```

My guess is that because you haven't specified the version, it is ignoring the `HintPath` and using the version in the GAC.

Try the following,

- I am on a Windows machine, so cannot check this, but can you add something like `Use specific version` to the NUnit reference in MonoDevelop?
- If that doesn't work, edit your project file to be like mine, but with your hint path.

Does that fix it?

**Comment by rprouse** [comment]
BTW, @mcraiha thanks for uploading your project file to pastebin. These things are much easier to diagnose when you can look at the code :+1:

**Comment by mcraiha** [comment]
@rprouse thanks. I tried Version addition, but it didn't help. I also removed old nunit from GAC, but error stays same. Error also stays same if Hintpath is bogus or if it is absolute path to .dll file.

Distro is 4.1.10-17.31.amzn1.x86_64 and Mono version is 4.3.0
I am compiling from terminal and I added NUnit by using nuget (**mono nuget.exe install NUnit** AND **mono nuget.exe install NUnit.Runners**) which installs 3.0.1 versions to home.

**Comment by rprouse** [comment]
@mcraiha, very odd. If you removed NUnit from the GAC, where is the old version coming from?

**Comment by CharliePoole** [comment]
This is now showing up in our own tests. See Travis build for PR #1128

**Comment by rprouse** [comment]
Looking at the callstack, I don't think it is because NUnit is the wrong version. I think it is likely the following code that is failing;

```C#
                case RuntimeType.Mono:
                    p.StartInfo.FileName = NUnitConfiguration.MonoExePath;
                    string monoOptions = "--runtime=v" + targetRuntime.ClrVersion.ToString(3);
                    if (debugTests || debugAgent) monoOptions += " --debug";
                    p.StartInfo.Arguments = string.Format("{0} \"{1}\" {2}", monoOptions, agentExePath, arglist);
                    break;
```

Specifically `targetRuntime.ClrVersion.ToString(3)`. If that is the case, then it is wrong. The docs on MSDN clearly states that "fieldCount ranges from 0 to 4."

It is also weird that this only hit our builds today. I wonder if a recent release of Mono has a bug in it. I will go check their source and report back.

**Comment by rprouse** [comment]
Interesting, the following code crashes with the same exception on Windows.

```C#
var vers = new Version(4, 5);
Console.WriteLine(vers.ToString(3));
```

As does `new Version("4.5")`

**Comment by rprouse** [comment]
@mcraiha I created a solution from your project and compiled it on Windows and it worked fine, but the `targetRuntime.ClrVersion` reports as v4.0.30319. My guess is that when compiled with the latest version of mono, it reports v4.5 which is the target in your csproj.

@CharliePoole, can you think of any reason why we can't just do `targetRuntime.ClrVersion.ToString()`? As far as I can tell, it reports correctly.

**Comment by CharliePoole** [comment]
We could try that but then we wouldn't know how it's getting set wrong in
the first place. The clr version of a Runtime Framework object is something
we set. It shouldn't be possible for it to be less than three components.

**Comment by rprouse** [comment]
I've done some experimenting with this on my Linux box and have been able to reproduce. @CharliePoole's issue in #922 is because the assembly being tested wasn't actually built and does not exist. Under Mono, it looks like we throw the ArgumentException if the assembly being tested does not exist. I tested on Windows and the correct File not found error is reported.

@mcraiha I noticed that you are passing in the CSPROJ file to test. We may have an additional error parsing the CSPROJ file, or setting the correct appbase directory, possibly only on Linux. Can you try running your tests by passing in the assembly instead of the project file?

To summarize, I think we might have two potential bugs here;

1. If the test assembly does not exist under Mono, we attempt to get the framework for the non-existing file, thus the exception.
2. Running tests using the csproj file does not work, possibly only on Linux?

**Comment by mcraiha** [comment]
@rprouse sorry about the delay. You are right, it works if I do

**xbuild myproject/nunit-lib.csproj**
**mono NUnit.Console.3.0.1/tools/nunit3-console.exe myproject/bin/Debug/nunit-lib.dll**

**Comment by CharliePoole** [comment]
@rprouse I think the two are related. What has been bothering me about this is that I didn't understand how we could get so deep in the tree of runners with a non-existent dll. I think it's possible that we are not checking for the existence of files under a project early enough, as we do when the assembly itself is entered on the command line.

**Comment by Oipo** [comment]
Just FYI, I just spend a couple hours figuring out why I got this exact same error on only one of my assemblies. I was trying to start the nunit3-console.exe on a misspelled filename, which obviously didn't exist. But that also triggers this bug.

**Comment by zmarty** [comment]
@Oipo: thanks, that was my problem as well.
